src/call/buffers.rs

```diff
@@ -0,0 +1,570 @@
+        }
+    }
+}
+
+impl Default for GrpcSlice {
+    fn default() -> Self {
+        GrpcSlice(unsafe { grpc_sys::grpc_empty_slice() })
+    }
+}
+
+impl Drop for GrpcSlice {
+    fn drop(&mut self) {
+        unsafe {
+            grpc_sys::grpcwrap_slice_unref(&self.0);
+        }
+    }
+}
+
+/// A wrapper for `grpc_byte_buffer_reader`.
+///
+/// Manages the `byte_buffer` behind the reader as well as the reader itself. That
+/// means we expect a 1:1 relationship between the `byte_buffer` and reader, so
+/// this is not a general purpose wrapper.
+struct GrpcByteBufferReader(grpc_byte_buffer_reader);
+
+impl GrpcByteBufferReader {
+    // Takes ownership of `buf` and will destroy it (at some point). The caller
+    // must not keep a reference.
+    //
+    // Safety: `buf` must be valid and non-null. The caller must not keep a reference
+    // to it.
+    unsafe fn new(buf: *mut grpc_byte_buffer) -> GrpcByteBufferReader {
+        let mut reader = MaybeUninit::uninit();
+        let init_result = grpc_sys::grpc_byte_buffer_reader_init(reader.as_mut_ptr(), buf);
+        assert_eq!(init_result, 1);
+        GrpcByteBufferReader(reader.assume_init())
+    }
+
+    fn len(&self) -> usize {
+        unsafe { grpc_sys::grpc_byte_buffer_length(self.0.buffer_out) }
+    }
+
+    fn next_slice(&mut self) -> GrpcSlice {
+        unsafe {
+            let mut slice = GrpcSlice::default();
+            let code = grpc_sys::grpc_byte_buffer_reader_next(&mut self.0, &mut slice.0);
+            debug_assert_ne!(code, 0);
+            slice
+        }
+    }
+}
+
+impl Drop for GrpcByteBufferReader {
+    fn drop(&mut self) {
+        unsafe {
+            let buf = self.0.buffer_in;
+            grpc_sys::grpc_byte_buffer_reader_destroy(&mut self.0);
+            grpc_sys::grpc_byte_buffer_destroy(buf);
+        }
+    }
+}
+
+/// Constructor function for a `grpc_slice` which wraps a Rust `Vec`.
+///
+/// Safety: see `VecSliceRefCount::new`, the same applies here.
+unsafe fn vec_slice(v: Vec<u8>) -> Box<grpc_slice> {
+    let mut data: MaybeUninit<grpc_sys::grpc_slice_grpc_slice_data> = MaybeUninit::zeroed();
+    *(*data.as_mut_ptr()).refcounted.as_mut() = grpc_sys::grpc_slice_grpc_slice_data_grpc_slice_refcounted {
+        bytes: v.as_ptr() as *const _ as *mut _,
+        length: v.len(),
+    };
+    let mut refcount = VecSliceRefCount::new(v);
+    let mut result = Box::new(grpc_slice {
+        refcount: ptr::null_mut(),
+        data: data.assume_init(),
+    });
+    refcount.slice = result.as_mut();
+    result.refcount = Box::into_raw(refcount) as *mut _;
+    result
+}
+
+/// A `grpc_slice_refcount` structure to handle ref-counting and memory mangement
+/// of `grpc_slice`s created by `vec_slice`.
+///
+/// Note that `grpc_slice_refcount` must be a prefix of this type so that a pointer
+/// to `VecSliceRefCount` can be treated polymorphically as a pointer to
+/// `grpc_slice_refcount`.
+#[repr(C)]
+struct VecSliceRefCount {
```

Comment:
I find the `vec_slice` is a little hard to understand. I suggest to reuse the struct `grpc_slice_refcount` to simplify the logic. That is struct `VecSliceRefCount` should look like
```
#[repr(C)]
struct VecSliceRefCount {
    grpc_slice_refcount rc;
    gpr_refcount refs;
    ptr: *mut u8,
    len: usize,
    cap: usize,
}
```

Reply:
We could replace the first two fields by a `grpc_slice_refcount`, and that would make the prefix requirement clearer. I don't understand the other fields. I think we must still keep a refcount and a pointer to the grpc_slice?

Reply:
I don't think we need to keep a pointer to grpc_slice, we just need to keep a reference to vec.

Reply:
`slice` is at a different address from the Vec. There is a pointer from the slice to the ref count object, but the destructor only gets a pointer to the ref count. So we need to keep both pointers so that we can drop both objects.

Reply:
I see. I think it's over complicated by introduce additional slice allocation. `grpc_slice` works like `bytes::Byte` somehow, it's unnecessary to put slice into heap.

Reply:
Hmm, if we put the `grpc_slice` inline in `VecSliceRefCount` that might simplify it. We can since the `VecSliceRefCount` must be heap allocated, we can safely return a pointer into it.